projects/Unit/TestAsyncConsumerExceptions.cs

```diff
@@ -0,0 +1,177 @@
+﻿// This source code is dual-licensed under the Apache License, version
+// 2.0, and the Mozilla Public License, version 2.0.
+//
+// The APL v2.0:
+//
+//---------------------------------------------------------------------------
+//   Copyright (c) 2007-2020 VMware, Inc.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       https://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//---------------------------------------------------------------------------
+//
+// The MPL v2.0:
+//
+//---------------------------------------------------------------------------
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+//  Copyright (c) 2007-2020 VMware, Inc.  All rights reserved.
+//---------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using NUnit.Framework;
+using RabbitMQ.Client.Events;
+
+namespace RabbitMQ.Client.Unit
+{
+    [TestFixture]
+    public class TestAsyncConsumerExceptions : IntegrationFixture
+    {
+        protected void TestExceptionHandlingWith(IBasicConsumer consumer,
+            Action<IModel, string, IBasicConsumer, string> action)
+        {
+            object o = new object();
+            bool notified = false;
+            string q = _model.QueueDeclare();
+
+
+            _model.CallbackException += (m, evt) =>
+            {
+                notified = true;
+                Monitor.PulseAll(o);
```

Comment:
Due to the sync call, the pulse will always happen before the wait, hence it always runs in the test timeout, but since its not checked in WaitOn-Method, it still passes but the test takes 5s.

My suggestion: Use a ResetEvent or a TaskCompletionSource to notify. (If you use a ResetEvent, you still have the timeout available)